src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java

```diff
@@ -93,4 +148,298 @@ public boolean checkSubAccountDetail(SubAccountGlobalDetail dtl) {
+        }
+        
+        success &= cgCostSharingEmptyOnGlobal;
+
+        return success;
+    }
+    
+    /**
+     * This method tests if all fields in the ICR section are empty.
+     *
+     * @return true if the ICR values passed in are empty, otherwise false.
+     */
+    protected boolean checkCgIcrIsEmpty() {
+        boolean success = true;
+        SubAccountGlobal newSubAccountGlobal = (SubAccountGlobal) super.getNewBo();
+        A21SubAccountChange newA21SubAccount = newSubAccountGlobal.getA21SubAccount();
+
+        if (ObjectUtils.isNotNull(newA21SubAccount)) {
+            success &= StringUtils.isEmpty(newA21SubAccount.getFinancialIcrSeriesIdentifier());
+            success &= StringUtils.isEmpty(newA21SubAccount.getIndirectCostRecoveryTypeCode());
+        }
+
+        return success;
+    }
+
+
+    /**
+     * This checks that if the cost share information is filled out that it is valid and exists, or if fields are missing (such as
+     * the chart of accounts code and account number) an error is recorded
+     * 
+     * @return true if all cost share fields filled out correctly, false if the chart of accounts code and account number for cost
+     *         share are missing
+     */
+    protected boolean checkCgCostSharingRules(String subAccountTypeCode) {
+
+        boolean success = true;
+        boolean allFieldsSet = false;
+        
+        SubAccountGlobal newSubAccountGlobal = (SubAccountGlobal) super.getNewBo();
+
+        A21SubAccountChange a21 = newSubAccountGlobal.getA21SubAccount();
+
+        // check to see if all required fields are set
+        if (StringUtils.isNotEmpty(a21.getCostShareChartOfAccountCode()) && StringUtils.isNotEmpty(a21.getCostShareSourceAccountNumber())) {
+            allFieldsSet = true;
+        }
+
+        // Cost Sharing COA Code and Cost Sharing Account Number are required
+        if (!allFieldsSet && StringUtils.isNotBlank(a21.getCostShareChartOfAccountCode())) {
+        	success &= checkEmptyBOField(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER, a21.getCostShareSourceAccountNumber(), getDisplayName(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT_NUMBER));
+        }
+        
+        if (!allFieldsSet && StringUtils.isNotBlank(a21.getCostShareSourceAccountNumber())) {
+        	success &= checkEmptyBOField(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_CHART_OF_ACCOUNTS_CODE, a21.getCostShareChartOfAccountCode(),  getDisplayName(KFSPropertyConstants.A21_SUB_ACCOUNT + "." + KFSPropertyConstants.COST_SHARE_SOURCE_CHART_OF_ACCOUNTS_CODE));
+        }
+
+        a21.refreshReferenceObject(KFSPropertyConstants.COST_SHARE_SOURCE_ACCOUNT);
+        // existence test on Cost Share Account
+        if (allFieldsSet) {
+            if (ObjectUtils.isNull(a21.getCostShareAccount())) {
```

Comment:
I see there are various places where the business rules are assuming the referenced objects exist if the FK references are valid. Is that necessarily the case here and in other places in these business rules? Are the reference objects guaranteed to be refreshed or reloaded at this stage?